packages/plugins/scripts-runner/src/index.ts

```diff
@@ -0,0 +1,223 @@
+import {
+  InitializationError,
+  UnsupportedTargetError,
+  ScriptExecutionError,
+  ScriptTrackingError
+} from './error';
+
+import { FileSystemTracker, ScriptsTracker, TrackingData } from './tracker';
+
+import AsyncIterator from 'core-js-pure/features/async-iterator';
+import { Embark, Callback } from 'embark-core';
+import { Logger } from 'embark-logger';
+import { dappPath } from 'embark-utils';
+import * as fs from 'fs-extra';
+import path from 'path';
+import Web3 from "web3";
+import { BlockTransactionObject } from 'web3-eth';
+
+const DEFAULT_TRACKING_FILE_PATH = '.embark/chains.json';
+
+export enum ScriptsRunnerCommand {
+  Initialize = 'scripts-runner:initialize',
+  Execute = 'scripts-runner:execute'
+}
+
+export enum ScriptsRunnerEvent {
+  Executed = 'scripts-runner:script:executed'
+}
+
+export interface ScriptsRunnerPluginOptions {
+  tracker?: ScriptsTracker;
+}
+
+export interface ScriptDependencies {
+  web3: Web3 | null;
+  contracts: any;
+  logger: Logger;
+}
+
+export interface ExecuteOptions {
+  target: string;
+  dependencies: ScriptDependencies;
+  forceTracking: boolean;
+}
+
+export default class ScriptsRunnerPlugin {
+
+  private _web3: Web3 | null = null;
+
+  private _block: BlockTransactionObject | null = null;
+
+  private trackingEnabled: boolean;
+
+  private tracker: ScriptsTracker;
+
+  constructor(private embark: Embark, options?: ScriptsRunnerPluginOptions) {
+    const trackingFilePath = dappPath(embark.config.contractsConfig?.tracking || DEFAULT_TRACKING_FILE_PATH);
+
+    this.tracker = options?.tracker ? options.tracker : new FileSystemTracker(trackingFilePath, embark.config.embarkConfig.migrations);
+    this.trackingEnabled = embark.config.contractsConfig.tracking !== false;
+
+    // TODO: it'd be wonderful if Embark called `registerCommandHandlers()` for us
+    this.registerCommandHandlers();
+  }
+
+  private registerCommandHandlers() {
+    this.embark.events.setCommandHandler(ScriptsRunnerCommand.Initialize, this.initialize.bind(this));
+    this.embark.events.setCommandHandler(ScriptsRunnerCommand.Execute, this.execute.bind(this));
+  }
+
+  private async initialize(callback: Callback<any>) {
+    if (!this.trackingEnabled) {
+      return callback();
+    }
+    try {
+      const web3 = await this.web3;
+      this.tracker.setWeb3(web3);
+      await this.tracker.ensureTrackingFile(this.embark.env);
+      callback();
```

Comment:
Here's another little thing to be aware of:

Notice that `web3` is being set in the tracker via `setWeb3()`. The reason this is done, again, boils down to us not having proper DI in place.

I ran into issues with testing the plugin because it tried to calculate a block hash for the tracker by calling a blockchain node. Unfortunately, there was no easy way to mock web3 in the plugin due to missing DI capabilities, so I've moved the hash calculation **into** the tracker. This then allowed me to bypass the web3 calls in a test mock tracker.

However, moving the block hash calculation into the tracker still means that it needs a web3 instance. Since web3 instances a retrieved asynchronous, they can as well, not be a passed to the tracker's constructor.

So requiring a `setWeb3()` API that is called as part of this initialization phase seemed the most flexible and testable option to me.

Reply:
I was considering that too, but then I thought, since the `FileSystemTracker` always comes with the `ScriptsRunnerPlugin` by default, it's probably not a good idea to duplicate the `web3` getter in multiple places.

But I can change that.